File: scraping/ksei/ksei.mjs
// fetch-ksei.mjs
import fetch from 'node-fetch';
import { writeFile } from 'fs/promises';
import { mkdir } from 'fs/promises';
import dotenv from 'dotenv';
import path from 'path';

dotenv.config();

const BASE_URL = 'https://akses.ksei.co.id/service/myportofolio/summary-detail';
const DATE = process.env.DATE || new Date(Date.now() - 2 * 24 * 60 * 60 * 1000).toISOString().split('T')[0];
const AUTH_TOKEN = process.env.AUTH_TOKEN;
const PORTFOLIO_DATA_PATH = process.env.PORTFOLIO_DATA_PATH || './data';

if (!AUTH_TOKEN) {
  console.error('❌ AUTH_TOKEN is not defined in environment variables.');
  process.exit(1);
}

const HEADERS = () => ({
  "accept": "*/*",
  "accept-language": "en-US,en-GB;q=0.9,en;q=0.8,id;q=0.7,eo;q=0.6",
  "authorization": AUTH_TOKEN,
  "sec-ch-ua": "\"Not)A;Brand\";v=\"8\", \"Chromium\";v=\"138\", \"Google Chrome\";v=\"138\"",
  "sec-ch-ua-mobile": "?0",
  "sec-ch-ua-platform": "\"Linux\"",
  "sec-fetch-dest": "empty",
  "sec-fetch-mode": "cors",
  "sec-fetch-site": "same-origin",
  "Referer": "https://akses.ksei.co.id/myportofolio/saldo"
});

const PORTFOLIO_TYPES = [
  'ekuitas',
  'reksadana',
  'obligasi',
  'kas',
  'lainnya'
];

async function fetchPortfolio(type) {
  const url = `${BASE_URL}/${type}?tanggal=${DATE}`;
  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: HEADERS(),
      credentials: 'include',
      mode: 'cors'
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status} (${response.statusText})`);
    }

    const data = await response.json();
    const filename = path.join(PORTFOLIO_DATA_PATH, `${type}.json`);
    await mkdir(PORTFOLIO_DATA_PATH, { recursive: true });
    await writeFile(filename, JSON.stringify(data, null, 2));
    console.log(`✅ [${type}] saved to ${filename}`);
  } catch (err) {
    console.error(`❌ [${type}] fetch failed: ${err.message}`);
  }
}

async function main() {
  await Promise.all(PORTFOLIO_TYPES.map(fetchPortfolio));
}

main();

File: scraping/ksei/ksei.py
import os
import asyncio
import aiohttp
import json
from datetime import datetime, timedelta
from pathlib import Path
from dotenv import load_dotenv
from fake_useragent import UserAgent

load_dotenv()
ua = UserAgent()

BASE_URL = 'https://akses.ksei.co.id/service/myportofolio/summary-detail'
DATE = os.getenv('DATE') or (datetime.now() - timedelta(days=2)).strftime('%Y-%m-%d')
AUTH_TOKEN = os.getenv('AUTH_TOKEN')
PORTFOLIO_DATA_PATH = Path(os.getenv('PORTFOLIO_DATA_PATH', './data'))

if not AUTH_TOKEN:
    print('❌ AUTH_TOKEN is not defined in environment variables.')
    exit(1)

HEADERS = {
    "authorization": AUTH_TOKEN,
    "Referer": "https://akses.ksei.co.id/myportofolio/saldo",
    "User-Agent": ua.random,
    "Content-Type": "application/json"
    
}

PORTFOLIO_TYPES = ['ekuitas', 'reksadana', 'obligasi', 'kas', 'lainnya']

async def fetch_portfolio(session, type_):
    url = f"{BASE_URL}/{type_}?tanggal={DATE}"
    try:
        async with session.get(url, headers=HEADERS) as response:
            if not response.ok:
                raise Exception(f"HTTP {response.status} ({response.reason})")
            
            data = await response.json()
            filename = PORTFOLIO_DATA_PATH / f"{type_}.json"
            PORTFOLIO_DATA_PATH.mkdir(parents=True, exist_ok=True)
            
            with open(filename, 'w') as f:
                json.dump(data, f, indent=2)
            
            print(f"✅ [{type_}] saved to {filename}")
    except Exception as e:
        print(f"❌ [{type_}] fetch failed: {e}")

async def main():
    async with aiohttp.ClientSession() as session:
        await asyncio.gather(*[fetch_portfolio(session, t) for t in PORTFOLIO_TYPES])

if __name__ == "__main__":
    asyncio.run(main())